docs/sdk-configuration.md

```diff
@@ -0,0 +1,492 @@
+inherit from a `ExporterBuilder` type of abstract class - it takes care of batching configuration to
+avoid having to understand separate concepts like `BatchSpanProcessor`. Exporters that don't support
+batching at all, like the custom `FullyAsyncZipkinExporter` don't have to. Note this can be
+similarly achieved by having wrappers like `BatchExporter.wrap(OtlpGrpcSpanExporter.builder()...`
+as well and is presented as one option.
+
+Note that parts can be separated to separate subconfig classes if it seems to make the API easier to
+use. We would want to avoid duplication though, e.g., having two configurations that expect a
+`Resource`.
+
+One thing to keep in mind is that it will be less common for an application developer to go this far
+and we can expect it is actually framework developers that use the full configuration capabilities
+of the SDK, likely by tying it to a separate configuration system.
+
+### Why build instances
+
+We have found that even at such an early stage of user adoption, users want to build instances of
+the SDK.
+
+- Integrates with dependency injection, e.g., Spring, in a similar way as many other libraries
+- Can allow having multiple instances in the same app, in multi-concern single-classloader scenarios
+- Allow managing lifecycle of SDK, e.g., shutting down and starting along with the lifecycle of a
+serverless runtime
+
+### Configuring the SDK in a framework
+
+It is extremely common for Java apps to be written using a dependency injection framework like
+Spring, Guice, Dagger, HK, and many more. They will all follow a very similar pattern though.
+
+```java
+@Component
+public class OpenTelemetryModule {
+  
+  @Bean
+  public OpenTelemetry openTelemetry(MonitoringConfig config) {
+    OpenTelemetrySdk.builder()
+      .setResource(Resource.getDefault().merge(CoolResource.getDefault()))
+      .setClock(AtomicClock.create())
+      .setPropagators(B3Propagator.getInstance())
+      .addSpanExporter(OtlpGrpcSpanExporter.builder()
+        .setEndpoint(config.getOtlpExporter().getEndpoint())
+        .setTimeout(config.getOtlpExporter().getTimeout())
+        .setBatchQueueSize(config.getOtlpExporter().getQueueSize())
+        .build())
+      .addSpanExporter(ZipkinExporter.builder()
+        .setEndpoint(config.getZipkinExporter().getEndpoint())
+        .disableBatching()
+        .build())
+      .addSpanExporter(FullyAsyncZipkinExporter.create())
+      .addSpanProcessor(CustomAttributeAddingProcessor.create())
+      .addSpanProcessor(CustomEventAddingProcessor.create())
+      .setTraceSampler(config.getSamplingRate() != 0 
+          ? Sampler.rateLimiting(config.getSamplingRate()) : Sampler.getDefault())
+      .setTraceLimits(TraceLimits.builder().setMaxAttributes(config.getMaxSpanAttributes()).build())
+      .setIdGenerator(TimestampedIdGenerator.create())
+      .setMeterStuff(InProgressMetricsStuff.create())
+      .build();
```

Comment:
Here is a sticking point that I see with the omnibus build-everything-at-the-OpenTelemetrySdk-level: For 1.0, we need to remove metrics from the OpenTelemetrySdk, which will mean that we *must* provide a way to build the pre-release MeterSdkProvider in a stand-alone manner. And, there may (probably will) be people who only want to use one of the signals, without the omnibus OpenTelemetrySdk, so we'll need to provide individual builders for each of the signals.

If we need to provide individual builders for each SDK signal, then are we ok with having both ways of building things? If so, then 👍 . This does seem to be a case where we should favor the end-user and make sure that they can do it either way.

Reply:
If we need to provide a way to configure and use just some of the signals, then I think that we should go this way:
> requiring the SDK components for each signal to be configured indivually, then have the OpenTelemetrySdk builder only compose the other components

The problem with supporting mix and match is this piece of code: https://github.com/open-telemetry/opentelemetry-java/pull/2231#issuecomment-741903911 .